app/Models/Question.php

```diff
@@ -0,0 +1,281 @@
+        }
+        if ($this->isOpen() || $this->isClosed()) {
+            throw new Exception("tried to open question when it has already been opened");
+        }
+        $this->update(['opened_at' => now()]);
+    }
+
+    /**
+     * Closes the question.
+     * @throws Exception if it has already been closed or if it is not even open.
+     */
+    public function close(): void
+    {
+        if ($this->isClosed()) {
+            throw new Exception("tried to close question when it has already been closed");
+        }
+        if (!$this->hasBeenOpened()) {
+            throw new Exception("tried to close question when it has not been opened");
+        }
+        $this->update(['closed_at' => now()]);
+    }
+
+    /**
+     * @return bool Whether the question is a multiple-choice question (with checkboxes).
+     */
+    public function isMultipleChoice(): bool
+    {
+        return $this->max_options > 1;
+    }
+
+    /**
+     * @param User $user
+     * @return bool Whether a certain user has already voted in the question.
+     */
+    public function hasVoted(User $user): bool
+    {
+        return $this->users()->where('users.id', $user->id)->exists();
+    }
+
+    /**
+     * Stores an answer (one or more options, or a long textual answer)
+     * for the question.
+     * If an answer sheet is provided,
+     * it also appends the answer to it.
+     * Throws if an option does not belong to the question,
+     * if too many options are selected,
+     * if the question is closed,
+     * if the user has already answered the question
+     * or if a long textual answer is provided for a question which does not support it.
+     */
+    public function storeAnswers(User $user, QuestionOption|array|string $answer, ?AnswerSheet $answerSheet = null): void
+    {
+        // the additional check is needed for the seeder
+        if (!$this->isOpen() && (!app()->runningInConsole() || app()->runningUnitTests())) {
+            throw new Exception("Tried to store answers for a question which is not open");
+        }
+
+        DB::transaction(function () use ($user, $answer, $answerSheet) {
+            // if we get only one option:
+            if ($answer instanceof QuestionOption) {
+                $answer = [$answer];
+            }
+
+            if (is_array($answer)) {
+                if (count($answer) > $this->max_options) {
+                    throw new Exception("More answers given then allowed ({$this->max_options})");
+                }
+                foreach ($answer as $option) {
+                    if ($option->question_id != $this->id) {
+                        throw new Exception("Received an option which does not belong to the question");
+                    }
+                    $option->increment('votes');
+
+                    if (isset($answerSheet)) {
+                        DB::table('answer_sheet_question_option')->insert([
```

Comment:
I might be misunderstanding this, but isn't it just a simple many to many relationship that could be managed by the ORM? I don't know whether it is recommended, but I would probably use `attach` or `detach` for it. https://laravel.com/docs/11.x/eloquent-relationships#updating-many-to-many-relationships

Reply:
The reason behind my suggestion is that inside Controllers, it should not matter how the data is managed inside the database and what are the field / table names as it should be handled inside the model.

Reply:
Oh, nvm, I think I can probably see why

Reply:
Yes; after the previous saga, I'm more inclined to manipulate the database directly :sweat_smile: